buildpacks/maven/src/main.rs

```diff
@@ -201,30 +209,51 @@ impl Buildpack for MavenBuildpack {
         // running since they might be confusing to the user.
         let internal_maven_options = vec![String::from("-B")];
 
-        log_header("Executing Maven");
-        log_info(format!(
-            "$ {} {} {}",
-            mvn_executable.to_string_lossy(),
-            shell_words::join(&maven_options),
-            shell_words::join(&maven_goals)
-        ));
-
-        util::run_command(
-            Command::new(&mvn_executable)
+        output::print_section("Running Maven build");
+        output::print_subsection(BuildpackOutputText::new(vec![
+            BuildpackOutputTextSection::regular("Running "),
+            BuildpackOutputTextSection::value(format!(
+                "{} {} {}",
+                mvn_executable.to_string_lossy(),
+                shell_words::join(&maven_options),
+                shell_words::join(&maven_goals)
+            )),
+        ]));
+
+        output::track_timing(|| {
+            let mut command = Command::new(&mvn_executable);
+
+            command
                 .current_dir(&context.app_dir)
                 .args(
                     maven_options
                         .iter()
                         .chain(&internal_maven_options)
                         .chain(&maven_goals),
                 )
-                .envs(&mvn_env),
-            MavenBuildpackError::MavenBuildIoError,
-            MavenBuildpackError::MavenBuildUnexpectedExitCode,
-        )?;
-
-        util::run_command(
-            Command::new(&mvn_executable)
+                .envs(&mvn_env);
+
+            output::run_command(
+                command,
+                false,
+                MavenBuildpackError::MavenBuildIoError,
+                |output| MavenBuildpackError::MavenBuildUnexpectedExitCode(output.status),
```

Comment:
I have some guidelines for running commands here https://github.com/heroku/buildpacks-ruby/pull/139. It seems that the output from the command being run won't be visible if this fails:

- [x] A representation of the command being run should appear in logs, or the error, or both
- [ ] Stdout and Stderr from the command should appear in logs, or the error, but NOT both (if it runs).
- [x] Make non-zero status an error
- [x] Allow for debugging a command that was run after execution (i.e., get command program name, arguments, env, cwd and pass them to something like WhichProblem)
- [x] Ideally only have to format Command once

Please update to include the output when this fails. I know some commands of yours produce absurd amount of data. If that's the case then maybe something like the last N lines of output? Basically what I'm after is: Why did this fail and what can I do about it? Having the status code is good, but I'm betting that there might be hints in the command output as well that will help.

Reply:
I'm not sure what the ask here is. Can you clarify? Currently, this outputs both `stdout` and `stderr` as it would on a local terminal. For Maven, the build error will be the last thing in the logs and Maven users are used to it. Are you suggesting adding the whole of `stderr` if the command fails, but filtering it out if it doesn't?

Reply:
> Currently, this outputs both stdout and stderr as it would on a local terminal. For Maven, the build error will be the last thing in the logs and Maven users are used to it. Are you suggesting adding the whole of stderr if the command fails, but filtering it out if it doesn't?

It's not outputting to stdout/stderr here. This code:

```rust
output::run_command(
    command,
    true,
    MavenBuildpackError::MavenBuildIoError,
    |output| MavenBuildpackError::MavenBuildUnexpectedExitCode(output.status),
)
```

Calls this branch where stdout and stderr are sent to `std::io::sink()` because quiet is true:

```
pub fn run_command<E, F: FnOnce(std::io::Error) -> E, F2: FnOnce(Output) -> E>(
    mut command: Command,
    quiet: bool,
    io_error_fn: F,
    exit_status_fn: F2,
) -> Result<Output, E> {
    let child = if quiet {
        command.output_and_write_streams(std::io::sink(), std::io::sink())
// ...
}
```

The docs for https://doc.rust-lang.org/beta/std/io/struct.Sink.html describe it as "A writer which will move data into the void." So it's never being seen by the user when quiet is set to true.

Reply:
I see the confusion now. I commented on the wrong line. The command invocation in question is L272, this one is clearly using the quiet = false argument. I had assumed the two maven commands had different error variants.

I'll close this comment and move it over to that line.